### DISCUSSION TITLE
Can't subscribe to a collection after authenticatation

### DISCUSSION THREAD

**Discussion by islamzaoui** [pocketbase/pocketbase#4598] [Q&A] [answered]
I have a problem with subscribing to collection after I authenticate with password until I refresh the page in my sveltekit app

I have a custom Hook for `onRealtimeAfterSubscribeRequest` and `onRealtimeDisconnectRequest` to log if user connected or not and it's doesn't display anything until I refresh the page

you can find my project [here](https://github.com/islamZaoui/groupchat), it's a simple public chat.

**Comment by ganigeorgiev** [comment]
It is difficult to understand the flow of your application and the problem you are describing, but my guess is that it is probably caused by the 2 JS SDK instances (client and server-side) not being in sync, for example the realtime subscribe call is being executed before the client SDK auth store is loaded with the node server-side auth cookie.

Try to add `console.log` statements before the `subscribe` calls to inspect the state of the `pb.authStore.token`.

_Side-note: You don't have to call `authRefresh` [here](https://github.com/IslamZaoui/GroupChat/blob/main/src/routes/%2Bpage.server.ts#L36C13-L36C63) as the `authWithPassword` call does the same - aka. on success it will populate the store with the auth token and user data that comes with the response._

**Reply by ganigeorgiev** [reply]
I haven't tested it, but note that your [hook.client.js](https://github.com/IslamZaoui/GroupChat/blob/main/src/hooks.client.ts) is probably executed only once and not on every page navigation, which is probably what you want since the server can modify the cookie between routes (to confirm that this is the case try to add a console.log in that file and check whether it is called after page navigation/redirect).

**Reply by ganigeorgiev** [reply, marked as answer]
If the latter turn out to be the problem, instead of manually invoking `loadFromCookie` on every client-side page file, an alterntive could be to make it computed/dynamic by creating a [custom store](https://github.com/pocketbase/js-sdk?tab=readme-ov-file#custom-auth-store) and call `document.cookie` from the getters. Something like (_I haven't tested it_):

```js
import { env } from '$env/dynamic/public';
import PocketBase, { BaseAuthStore, cookieParse } from 'pocketbase';

class CookieStore extends BaseAuthStore { // extends BaseAuthStore but if you need the data to be also in localStorge then replace with LocalAuthStore
    constructor() {
        super();
        
        this.loadFromCookie(document.cookie)
    }

    get token() {
        try {
            const data = JSON.parse(cookieParse(document.cookie)['pb_auth']);
            return data.token
        } catch (_) {}

        return super.token;
    }

    get model() {
        try {
            const data = JSON.parse(cookieParse(document.cookie)['pb_auth']);
            return data.model
        } catch (_) {}

        return super.model;
    }

    save(token, model) {
        super.save(token, model)

        document.cookie = this.exportToCookie({ httpOnly: false })
    }

    clear() {
        super.clear()

        document.cookie = this.exportToCookie({ httpOnly: false })
    }
}

export const pb = new PocketBase(env.PUBLIC_POCKETBASE_URL, new CookieStore());
```

**Reply by islamzaoui** [reply]
[hook.client.js](https://github.com/IslamZaoui/GroupChat/blob/main/src/hooks.client.ts) is executing after every request made so it's not the problem, but the client side hooks aren't on sync with server, I tried your solution and it's working after I did small modification
```ts
import { browser } from '$app/environment';
import { env } from '$env/dynamic/public';
import PocketBase, { BaseAuthStore, cookieParse, type AuthModel } from 'pocketbase'

class CookieStore extends BaseAuthStore {
  constructor() {
    super();
    /* I add this condition to fix issue with SSR on sveltekit */
    if (browser)
      this.loadFromCookie(document.cookie)
  }

  get token() {
    if (browser)
      try {
        const data = JSON.parse(cookieParse(document.cookie)['pb_auth']);
        return data.token
      } catch (_) { }

    return super.token;
  }

  get model() {
    if (browser)
      try {
        const data = JSON.parse(cookieParse(document.cookie)['pb_auth']);
        return data.model
      } catch (_) { }

    return super.model;
  }

  save(token: string, model: AuthModel | undefined) {
    super.save(token, model)

    if (browser)
      document.cookie = this.exportToCookie({ httpOnly: false })
  }

  clear() {
    super.clear()

    if (browser)
      document.cookie = this.exportToCookie({ httpOnly: false })
  }
}

export function createInstance(store?: CookieStore) {
  return new PocketBase(env.PUBLIC_POCKETBASE_URL, store)
}

export const pb = createInstance(new CookieStore())
```

there is another option is to load cookies in onMount, afterUpdate and beforeNavigation in the root layout.svelte

thanks for helping me and have a nice day
